File: src/ledger.ts
import { EvidenceLedger, Statute } from './types';

/** The set of citation ids the gate will accept: every fact id + every statute id. */
export function validCitationIds(ledger: EvidenceLedger, statutes: Statute[]): Set<string> {
  const ids = new Set<string>();
  for (const f of ledger.facts) ids.add(f.id);
  for (const s of statutes) ids.add(s.id);
  return ids;
}

/** Compact, model-friendly rendering of the ledger for inclusion in prompts. */
export function renderLedger(ledger: EvidenceLedger): string {
  return ledger.facts
    .map((f) => `[${f.id}] ${f.statement}  (source: ${f.source}; confidence ${f.confidence})`)
    .join('\n');
}

export function renderStatutes(statutes: Statute[]): string {
  return statutes
    .map((s) => `[${s.id}] ${s.title} (${s.jurisdiction})\n"""${s.text.trim()}"""`)
    .join('\n\n');
}

File: backend/app/ledger.py
"""Ledger / statute rendering + the set of valid citation ids (mirrors src/ledger.ts)."""
from __future__ import annotations

from .types import EvidenceLedger, Statute


def valid_citation_ids(ledger: EvidenceLedger, statutes: list[Statute]) -> set[str]:
    ids = {f.id for f in ledger.facts}
    ids.update(s.id for s in statutes)
    return ids


def render_ledger(ledger: EvidenceLedger) -> str:
    return "\n".join(
        f"[{f.id}] {f.statement}  (source: {f.source}; confidence {f.confidence})" for f in ledger.facts
    )


def render_statutes(statutes: list[Statute]) -> str:
    return "\n\n".join(f'[{s.id}] {s.title} ({s.jurisdiction})\n"""{s.text.strip()}"""' for s in statutes)
